### DISCUSSION TITLE
How do I test Python programs with signal handlers (e.g., SIGUSR1, SIGINT, etc.)?

### DISCUSSION THREAD

**Discussion by achimnol** [community/community#25663] [Actions]
<p>Running signal-based unit tests makes the GitHub Actions to hang or terminate prematurely.<br>
The test suite works perfectly fine on local environments, specifically, when executed from a shell prompt.<br>
Is there any way to let my test suite to run as a separate process group or session, so that intentionally raised signals are delivered to only the test program, not the action runner?</p>

**Comment by BrightRan** [comment]
<p><a class="mention" href="/achimnol">@achimnol</a>,</p>
<p>On GitHub Actions workflow, we seem have no available easy methods to do that. There are not any actions for signal-based unit tests can be found on the <a href="https://github.com/marketplace" rel="noopener nofollow ugc"><strong>GitHub Marketplace</strong></a>.</p>

**Comment by Simran-B** [comment]
<p>Maybe there’s a different way to test signal handling?</p>
<aside class="onebox stackexchange">
  <header class="source">
      <a href="https://stackoverflow.com/questions/26158373/how-to-really-test-signal-handling-in-python" target="_blank" rel="noopener">stackoverflow.com</a>
  </header>
  <article class="onebox-body">
      <a href="https://stackoverflow.com/users/759076/jackson-tale" target="_blank" rel="noopener">
    <img alt="Jackson Tale" src="https://user-images.githubusercontent.com/7819991/181092418-9446aeae-492a-438b-bf02-ec4ca15d57f8.jpeg" class="thumbnail onebox-avatar" width="128" height="128">
  </a>
<h4>
  <a href="https://stackoverflow.com/questions/26158373/how-to-really-test-signal-handling-in-python" target="_blank" rel="noopener">How to really test signal handling in Python?</a>
</h4>

<div class="tags">
  <strong>python, unit-testing, signals</strong>
</div>

<div class="date">
  asked by
  
  <a href="https://stackoverflow.com/users/759076/jackson-tale" target="_blank" rel="noopener">
    Jackson Tale
  </a>
  on <a href="https://stackoverflow.com/questions/26158373/how-to-really-test-signal-handling-in-python" target="_blank" rel="noopener">09:51AM - 02 Oct 14 UTC</a>
</div>

  </article>
  <div class="onebox-metadata">
    
    
  </div>
  <div style="clear: both"></div>
</aside>

**Comment by achimnol** [comment, marked as answer]
<p>I managed to resolve the issue by calling <code>os.setpgrp()</code> to split the signal-receiving process group in the test host process.</p>
